static_frame/core/type_blocks.py

```diff
@@ -2612,6 +2611,9 @@ def drop(self, key: GetItemKeyTypeCompound) -> 'TypeBlocks':
                 shape_reference=self._shape
                 )
 
+    def __iter__(self) -> tp.Iterator['TypeBlocks']:
+        for i in range(self.shape[1]):
+            yield self._extract(column_key=i)
```

Comment:
I see. I dislike added code just to make MyPy happy, and in this case, as I point out, there are two sensible options here as to whether an `np.ndarray` or a `TypeBlocks` instance should be extracted. But if Python simply does the same thing under `iter` without implementing this, I favor not implementing this function and silencing MyPy's complaint.

Reply:
Isn't it better to be explicit though, instead of relying on default (and potentially confusing & opaque) behavior?

Or maybe to combine with your initial suggestion, should we make `__iter__`'s defintion raise `NotImplementedError`, and instead make an explicit method called `iter_array`?

Reply:
Raising `NotImplemented()` is attractive. I do not think we need an `iter_array`, so perhaps nothing more to do at this point.

Reply:
What do you mean we mean we don't need `iter_array`? There are multiple places throughout the code base that `for block in blocks` is called.

If we don't want to define `__iter__` for the ambiguity you've mentioned, I don't know how not defining it, but relying on the default implementation addresses the issue of "What does `for block in blocks` even mean?"

My though is to set `__iter__` to raise `NotImplementedError` (not `NotImplemented`), and instead replace all instances throughout the code base with `for block in blocks.iter_array()`.

Reply:
Are you sure that, in the places we are calling `for block in blocks`, `blocks` is a `TypeBlocks`? What I see as far more common is that `blocks`, in those scenarios, is a `tp.Iterable[np.ndarray]` (which we often refer to as "blocks"), not `TypeBlocks`.

I am not opposed to implementing `iter_array`, I just suspect we do not need it. If you have a case in the `IndexHierarchy` implementation, then that might be good enough reason.

And yes, `NotImplementedError` not `NotImplemented`!

Reply:
Oooooo I see now. Okay yeah no one is actually iterating over a `TypeBlocks` instance. I will raise an error, and just implement something explicit in `IndexHierarchy`